k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+ *  @endcode
+ *
+ *  @code{.cpp}
+ *  K2_ASSERT(1 == 1);
+ *  @endcode
+ */
+#ifndef K2_ASSERT
+#define K2_ASSERT(exp) assert(exp)
+#endif
+
+/**
+ * @brief Check if two arguments are equal.
+ *
+ * @details Implemented by `assert`, which is supported by host and device.
+ *
+ * @param[in] a left argument to compare
+ * @param[in] b right argument to compare
+ *
+ * @note
+ *  Assertions are for debugging purposes. They can affect performance
+ *  and it is therefore recommended to disable them in production code.
+ *  `assert` can be disabled at compile time by defining the `NDEBUG`
+ *  preprocessor macro before including assert.h.
+ *  Thus, `K2_MAKE_ERROR_CHECK` is ignored here.
+ *
+ *  @code{.cpp}
+ *  K2_CHECK_EQ(1, 1);
+ *  @endcode
+ */
+#ifndef K2_CHECK_EQ
+#define K2_CHECK_EQ(a, b) assert( a == b )
+#endif
+
+/**
+ * @fn
+ *  __host__ __device__ __forceinline__ cudaError_t
+ *  _K2CudaDebug(cudaError_t error,
+ *               const char *filename,
+ *               int line)
+ *
+ *
+ * @brief This is a error checking function, with context information.
+ *        It's not designed to called by users, but inner macros.
+ *
+ * @details
+ *  If K2_MAKE_ERROR_CHECK is defined and error is not cudaSuccess,
+ *  the corresponding error message is printed to:
+ *    - host: stderr
+ *    - device: stdout in device
+ *  along with the supplied source context.
+ *  It's used to make other macros convenient.
+ *
+ * @param[in] error         an enum type indicating CUDA errors.
+ * @param[in] filename      the source filename that the error comes from.
+ * @param[in] line          the code line that the error happened.
+ * @return                  Pass the input CUDA error.
+ *
+ * @code{.cpp}
+ * _K2CudaDebug(cudaGetLastError(), __FILE__, __LINE__);
+ * @endcode
+ */
+__host__ __device__ __forceinline__ cudaError_t _K2CudaDebug(
+    cudaError_t error,
+    const char *filename,
+    int line) {
+#ifdef K2_MAKE_ERROR_CHECK
+  (void)filename;
+  (void)line;
+  if (cudaSuccess != error) {
+  #if (K2_PTX_ARCH == 0)
+    fprintf(stderr, "CUDA error ID=%d, NAME=%s, [%s, %d]: %s\n",
+            error, cudaGetErrorName(error),
+            filename, line,
+            cudaGetErrorString(error));
+    fflush(stderr);
+  #elif (K2_PTX_ARCH >= 200)
+    printf("CUDA error ID=%d, NAME=%s, "
+           "[block (%d,%d,%d) thread (%d,%d,%d), %s, %d]: %s\n",
+           error, cudaGetErrorName(error),
+           blockIdx.x, blockIdx.y, blockIdx.z,
+           threadIdx.x, threadIdx.y, threadIdx.z,
+           filename, line,
+           cudaGetErrorString(error));
+  #endif
+  }
+#endif
+  return error;
+}
+
+/**
+ * @def K2_CUDA_CHECK_ERROR(cudaError)
+ *
+ * @brief Macro for checking cuda error.
+ *
+ * @details
+ *  If `K2_MAKE_ERROR_CHECK` is defined and e is not cudaSuccess,
+ *  print the error message and exit with the error enum value.
+ *  Otherwise, it does nothing.
+ *
+ * @param[in] e an enum type indicating CUDA errors.
```

Comment:
use the type please, e.g. cudaError_t.  There are many such enum's in CUDA and they are not all the same.

Reply:
.. I meant, mention the type cudaError_t when you document the variable